# Allele Frequency Tools  

## Introduction of the tool  

This tool is used as a "blueprint" of all variants that found in the in-house uploaded sequencing samples. The webpage mainly including two sections:
- The **SNPs** in-house allele frequencies that dynamically calculated from all the uploaded samples.
- The **CNV/SVs** searching functions, that allow users to find a specific genomic regions, checking the CNVs/SVs records that existed in the uploaded samples.

## Quick start

### SNP allele frequencies
<img width="1197" alt="image" src="https://github.com/user-attachments/assets/a0303c2b-93d6-4a04-a6a0-0a6cfd5cfc98" />

All of SNPs are named as the format chr:pos:Ref/Alt, the percentage indicated the newest calculated allele frequency. 

1. **View files**  
By clicking this button, a list of all uploaded files (by sample name) are shown.  
  
2. **Export all calculation results of in-house allele frequencies**  
The calculation is based on the formula:
Allele Frequency = Total allele count of the SNP / 2* (Total Number of Individuals) whereas Total allele count is calculated referring the heterozygous/homozygous status of the variant.  

3. **Search / Scrolling tab**  
Here, by searching (input the chromosome location of the variants that interested) and scrolling the user is able to view all variants detected.

4. **Expand/More details**
Clicking this arrow will show all the sample IDs that carrying the specific variant being searched.

5. **Export variant-specific results**   
This button allows downloading of a CSV file including all the sample IDs that carrying the specific variant being searched.  

6. **View specific raw results table**  
This button allows expanding the raw result file (generated by GenRiskPro/BabySeqPro,including prioritized variants list with gene/disease information) for the specific sample.

<img width="1197" alt="image" src="https://github.com/user-attachments/assets/bb32eb85-7192-4dc6-bfa6-541f27e78c9f" />

7. **Back to all samples list**  
In the result file page, clicking "Back" will returns to the list of all samples uploaded.  

8. **Export whole sample-specific Results (CSV format)**  
This button allows downloading a CSV file of raw result file of the specific sample.  

<img width="1199" alt="image" src="https://github.com/user-attachments/assets/2204ecae-714d-4468-ab8c-f9aaac48e65d" />

9. **Sample-based view page**  
From the button **1** or button **7**, users will return to a page showing all samples that uploaded. The search tab allows searching by sample IDs as well as scrolling by mouse.  


---
### CNVs/SVs investigation based on DECIPHER database
<img width="1197" alt="image" src="https://github.com/user-attachments/assets/4458a96a-8810-4613-9ce5-a4effa740842" />  

1. **The search tab for the genes/regions of interest**  
You can search your interested genes/regions here.

2. **Quick examples of searching formats**    
Here, we provided 3 different ways of searching: 1) by gene name 2) by band information and 3) by chromosome positions. Clicking any of them will automatically applies the example search and showing the searching results.

3. **Checking CNV/SVs records from DECIPHER**
After searching, the page showed the data that retrived from DECIPHER database (CNVSVs section). Each bar represents a known insertion (red) or deletion (blue) on that region.By clicking the bar it will showing the details information of the CNV/SVs as shown in the figure below.

4. **Retriving co-located CNV/SVs from uploaded data**  
This "Run" button will retriving all co-located CNVs/SVs inside the range of the selected "bar" (CNV/SVs).The results were prioritized and re-ranked based on the criteria: 
Ranks variants by how “close” or how extensively they overlap with the queried range in basepairs. For example, if you’re looking in the range 1–10 and there is:

- Deletion1 in 4–5  
- Deletion2 in 2–8  

Then **Deletion2** would have a higher priority because it overlaps (or covers) more of the target range (1–10). In other words, a small deletion within a larger range gets a lower rank compared to a deletion that more fully spans or intersects that same range.

5. **Export matching Results (CSV format)**   
The page showed can be downloaded as CSV format files.  

6. **The queried range can be reduced by modifying button**  
The tool will then apply the same overlap logic but within the newly specified boundaries.

<img width="1197" alt="image" src="https://github.com/user-attachments/assets/e9ab0113-6bca-49f9-bf96-e255892a9bcd" />

