Once the job finishes, a `done` is marked on the job. Results could be viewed in the portal either in `PDF View` or `Raw Data`.



## **PDF View**

Results with clinical importance are formatted into the PDF report and previewed in the portal. The report contains the following contents:

1. Understanding Your Report: Definition of different medical impact labels and descriptions of eleven disease groups. Medical impacts are filtered by the ClinVar pathogenicity levels, where only pathogenic/likely pathogenic variants and variants with conflicting classifications are included in the PDF report. Predicted loss of function variants which has high impact on translated proteins are also included in the PDF report, classified as “For Your Information”.
2. Summary of Health Results: Variant information related to ACMG secondary findings v3.2 list, disease of carrier status, and pharmacogenetics (PGx) are briefly summarised in this section. Variants are grouped by medical impact levels and disease groups. PGx results are grouped by medical impact levels and phenotype category.
3. Detailed Health Results: Detailed information of the reported variants.
4. Traits: Detailed descriptions of the traits and every specific relevant genotype are illustrated in this section.
5. Technical Note: This section attaches the background gene list that is used in the current pipeline analysis.

The PDF report can be exported by clicking the exporting button on the top right corner and saved to local.

![image](https://github.com/user-attachments/assets/ce92062d-616c-43dc-b581-920ce92abeee)




## **Raw Data**

Variant selection and filtration could be applied with the Raw Data function. In the `Whole Genome` group, all screened diseases at risk could be found. Variants related to specific gene-disease lists (ACMD secondary findings, newborn screening list, carrier screening list, and cancer predisposition) could be viewed in each group.

Additional information contains all detailed results for a specific gene/variant. Content could be hidden by switching the hidden button. Hidden genes/variants will be not exported in the Excel file, but can be found in the `Hidden Genes/Variants` in the bottom right corner and can be regenerated back to the result list.

Genome, traits, and PGx results are separated and contents could be switched by selecting the group.

The filtration section provides genes or gene list filtration. Results related to the selected genes will be displayed in the portal.

Eventually, all data are able to export in `.xlsx` format to local.

![image](https://github.com/user-attachments/assets/87b59e5d-5364-4ad8-bafd-462277ed0eb4)
